src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+            return self;
+        }
+
+        @Specialization(guards = { "!isNil(options)", "wasProvided(options)" })
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyBasicObject options) {
+            return self;
+        }
+    }
+
+    @CoreMethod(names = "initialize_copy", required = 1)
+    public abstract static class InitializeCopyNode extends CoreMethodArrayArgumentsNode {
+        @Specialization
+        protected RubyConcurrentHashMap initializeCopy(RubyConcurrentHashMap self, RubyConcurrentHashMap other) {
+            self.concurrentHash.putAll(self.concurrentHash);
+            return self;
+        }
+    }
+
+    @CoreMethod(names = "[]", required = 1)
+    public abstract static class GetIndexNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object getIndex(RubyConcurrentHashMap self, Object key) {
+            final Object value = self.concurrentHash.get(new RubyConcurrentHashMap.Key(key));
+            if (value == null) {
+                return nil;
+            } else {
+                return value;
+            }
+        }
+    }
+
+    @CoreMethod(names = "[]=", required = 2)
+    public abstract static class SetIndexNode extends CoreMethodArrayArgumentsNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object setIndex(RubyConcurrentHashMap self, Object key, Object value) {
+            self.concurrentHash.put(new RubyConcurrentHashMap.Key(key), value);
+            return value;
+        }
+    }
+
+    @CoreMethod(names = "compute_if_absent", required = 1, needsBlock = true)
+    public abstract static class ComputeIfAbsentNode extends YieldingCoreMethodNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object computeIfAbsent(RubyConcurrentHashMap self, Object key, RubyProc block) {
+            Object returnValue = self.concurrentHash.computeIfAbsent(new RubyConcurrentHashMap.Key(key), k -> {
+                return callBlock(block);
+            });
+
+            if (returnValue == null) {
+                return nil;
+            }
+            return returnValue;
+        }
+    }
+
+    @CoreMethod(names = "compute_if_present", required = 1, needsBlock = true)
+    public abstract static class ComputeIfPresentNode extends YieldingCoreMethodNode {
+        @TruffleBoundary
+        @Specialization
+        protected Object computeIfPresent(RubyConcurrentHashMap self, Object key, RubyProc block) {
+            Object returnValue = self.concurrentHash.computeIfPresent(new RubyConcurrentHashMap.Key(key), (k, v) -> {
+                Object newValue = callBlock(block, v);
+
+                if (newValue == nil) {
+                    return null;
+                } else {
+                    return newValue;
+                }
+            });
+
+            if (returnValue == null) {
```

Comment:
Here we do need the check (as it is), computeIfPresent() is not really symmetrical with computeIfAbsent even though it sounds like it.

Reply:
> Gah I've just realised that some of these methods compare _values_, like `replace_pair`. This is really subtle. @wildmaples if you keep going as you are I will spec and fix this bug and explain.

This is fixed.